### ISSUE TITLE
Proposal to develop official "Developers' Guidelines"

### ISSUE THREAD

**Issue by flimzy** [gopherjs/gopherjs#331]
I'd like to suggest that it might be worth while to begin formulating a sort of **Developers' Guidelines** document somewhere, which documents official and semi-official guidelines, preferences, and expectations of new developers contributing to GopherJS directly, and perhaps those writing modules for use by the GopherJS community.

As a recent sample of two items that might fit in such a document:

 - [Stance on ECMAScript 6](https://github.com/gopherjs/gopherjs/issues/320)
 - [Always document exported functions, and otherwise follow Go upstream standards](https://github.com/gopherjs/gopherjs/pull/328)

Also (added by @shurcooL):

- [Proposal: Move towards a more flexible, low level compiler testing system like that of Go.] (https://github.com/gopherjs/gopherjs/issues/284)

I realize there are likely very few "official" decisions to document right now, but I expect that will change (and is changing!).

Currently, we have just the issue tracker, but something more digested for newcomers I think would be appropriate.

Would this be welcomed?  And if so, where should it live? Some obvious (to me) options:

 - DEVELOPERS.md in the main repo
- On the wiki
- The [gopherjs-book](https://github.com/gopherjs/gopherjs-book) repo (??)

**Comment by traveller42** [comment]
I feel some kind of Developers' Guide is useful.  Location is less important than ensuring it is easy to find.

It should include information on how to find previous discussion on an issue.  One doesn't want to spend time on something that the wider community has either rejected or deferred for reasons that may not be obvious to the newcomer.

A style guide is important in keeping the look of the code consistent.  I will submit that once such is in place, the existing code will need to be reviewed to provide a good example for comparison.

I think the links to the earlier items as above are good and should be encouraged.

**Comment by shurcooL** [comment]
I agree, this would be useful to as a place to collect and post this information, that is currently mostly scattered throughout closed and open issues.

I think the Wiki is a good place to put this.

We can start by adding Richard's [comment](https://github.com/gopherjs/gopherjs/pull/328#discussion_r42315597), and the other issues you've linked.

**Comment by flimzy** [comment]
I put some (very minimal) information on the wiki to get this project started: https://github.com/gopherjs/gopherjs/wiki/Developer-Guidelines

If we're happy with this, we can probably close this issue, and just start keeping that page up to date.

**Comment by shurcooL** [comment]
That looks like a great start, thanks for doing it!

**Comment by shurcooL** [comment]
> If we're happy with this, we can probably close this issue, and just start keeping that page up to date.

Let's do that. I think that page looks great and it's useful. Thanks again for starting this!
